wenet/utils/executor_nst.py

```diff
@@ -0,0 +1,169 @@
+
+                num_seen_utts += num_utts
+                if batch_idx % accum_grad == 0:
+                    if rank == 0 and writer is not None:
+                        writer.add_scalar('train_loss', loss, self.step)
+                    # Use mixed precision training
+                    if use_amp:
+                        scaler.unscale_(optimizer)
+                        grad_norm = clip_grad_norm_(model.parameters(), clip)
+                        # Must invoke scaler.update() if unscale_() is used in
+                        # the iteration to avoid the following error:
+                        #   RuntimeError: unscale_() has already been called
+                        #   on this optimizer since the last update().
+                        # We don't check grad here since that if the gradient
+                        # has inf/nan values, scaler.step will skip
+                        # optimizer.step().
+                        scaler.step(optimizer)
+                        scaler.update()
+                    else:
+                        grad_norm = clip_grad_norm_(model.parameters(), clip)
+                        if torch.isfinite(grad_norm):
+                            optimizer.step()
+                    optimizer.zero_grad()
+                    scheduler.step()
+                    self.step += 1
+                if batch_idx % log_interval == 0:
+                    lr = optimizer.param_groups[0]['lr']
+                    log_str = 'TRAIN Batch {}/{} loss {:.6f} '.format(
+                        epoch, batch_idx,
+                        loss.item() * accum_grad)
+                    if loss_att is not None:
+                        log_str += 'loss_att {:.6f} '.format(loss_att.item())
+                    if loss_ctc is not None:
+                        log_str += 'loss_ctc {:.6f} '.format(loss_ctc.item())
+                    log_str += 'lr {:.8f} rank {}'.format(lr, rank)
+                    logging.debug(log_str)
+
+    def cv(self, model, data_loader, device, args):
+        ''' Cross validation on
+        '''
+        model.eval()
+        rank = args.get('rank', 0)
+        epoch = args.get('epoch', 0)
+        log_interval = args.get('log_interval', 10)
+        # in order to avoid division by 0
+        num_seen_utts = 1
+        total_loss = 0.0
+        with torch.no_grad():
+            for batch_idx, batch in enumerate(data_loader):
+                key, feats, target, feats_lengths, target_lengths = batch
+                feats = feats.to(device)
+                target = target.to(device)
+                feats_lengths = feats_lengths.to(device)
+                target_lengths = target_lengths.to(device)
+                num_utts = target_lengths.size(0)
+                if num_utts == 0:
+                    continue
+                loss, loss_att, loss_ctc = model(feats, feats_lengths, target,
+                                                 target_lengths)
```

Comment:
We have changed the return type from multiple tensors to loss_dict, plz see https://github.com/wenet-e2e/wenet/pull/1261/files#r909610210